File: September2022/C++/CountGoodNodesInBinaryTree.cpp
class Solution {
private:
    int helper(TreeNode* root, int currMax) {
        if (root == NULL) {
            return 0;
        }
        int count = 0;
        if (root->val >= currMax) {
            count += 1;
            currMax = root->val;
        }
        count += helper(root->left, currMax);
        count += helper(root->right, currMax);
        return count;
    }
    
public:
    // TC: O(n)
    // SC: O(h)
    int goodNodes(TreeNode* root) {
        return helper(root, root->val);
    }
};

File: September2022/Java/CountGoodNodesInBinaryTree.java

// @saorav21994
// TC : O(n)
// SC : O(n) -> considering recursion stack space

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int res = 0;
    public int goodNodes(TreeNode root) {
        
        dfs(root, root.val);
        return res;
    }
    
    private void dfs(TreeNode node, int max) {
        if (node == null)
            return;
        if (node.val >= max) {
            max = node.val;
            res += 1;
        }
        if (node.left != null)
            dfs(node.left, max);
        if (node.right != null)
            dfs(node.right, max);
        
    }
}
